Mathlib/AlgebraicGeometry/Morphisms/AffineAnd.lean

```diff
@@ -0,0 +1,184 @@
+        simp_rw [Scheme.preimage_basicOpen] at hf
+        exact (hf ⟨r, hr⟩).left
+    refine ⟨inferInstance, hQs.ofIsLocalization hPi (f.app ⊤) s hs fun a ↦ ?_⟩
+    refine ⟨Γ(Y, Y.basicOpen a.val), Γ(X, X.basicOpen (f.app ⊤ a.val)), inferInstance,
+      inferInstance, inferInstance, inferInstance, inferInstance, ?_, ?_⟩
+    · exact (isAffineOpen_top X).isLocalization_basicOpen (f.app ⊤ a.val)
+    · obtain ⟨_, hf⟩ := hf a
+      rw [morphismRestrict_app, hPi.cancel_right_isIso, Scheme.Opens.ι_image_top] at hf
+      rw [(isAffineOpen_top Y).app_basicOpen_eq_away_map _ (isAffineOpen_top X)] at hf
+      rwa [hPi.cancel_right_isIso] at hf
+
+/-- If `P` is stable under base change, so is `affineAnd P`. -/
+lemma affineAnd_stableUnderBaseChange (hQi : RingHom.RespectsIso Q)
+    (hQb : RingHom.StableUnderBaseChange Q) :
+    (affineAnd Q).StableUnderBaseChange := by
+  haveI : (affineAnd Q).toProperty.RespectsIso := affineAnd_respectsIso hQi
+  apply AffineTargetMorphismProperty.StableUnderBaseChange.mk
+  intro X Y S _ _ f g ⟨hY, hg⟩
+  exact ⟨inferInstance, hQb.pullback_fst_app_top _ hQi f _ hg⟩
+
+lemma targetAffineLocally_affineAnd_iff (hQi : RingHom.RespectsIso Q)
+    {X Y : Scheme.{u}} (f : X ⟶ Y) :
+    targetAffineLocally (affineAnd Q) f ↔ ∀ (U : Y.affineOpens),
+      IsAffineOpen (f ⁻¹ᵁ U) ∧ Q (f.app U) := by
+  simp only [targetAffineLocally, affineAnd_apply, morphismRestrict_app, hQi.cancel_right_isIso]
+  refine ⟨fun hf U ↦ ?_, fun h U ↦ ?_⟩
+  · obtain ⟨hfU, hf⟩ := hf U
+    exact ⟨hfU, by rwa [Scheme.Opens.ι_image_top] at hf⟩
+  · refine ⟨(h U).1, ?_⟩
+    rw [Scheme.Opens.ι_image_top]
+    exact (h U).2
+
+end
+
+section
+
+variable {Q : ∀ {R S : Type u} [CommRing R] [CommRing S], (R →+* S) → Prop}
+
+/-- If `P` is a morphism property affine locally defined by `affineAnd Q`, `P` is stable under
+composition if `Q` is. -/
+lemma HasAffineProperty.affineAnd_isStableUnderComposition {P : MorphismProperty Scheme.{u}}
+    [HasAffineProperty P (affineAnd Q)] (hQ : RingHom.StableUnderComposition Q) :
+    P.IsStableUnderComposition where
+  comp_mem {X Y Z} f g hf hg := by
+    wlog hZ : IsAffine Z
+    · rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := P) _ (iSup_affineOpens_eq_top _)]
+      intro U
+      rw [morphismRestrict_comp]
+      exact this hQ _ _ (IsLocalAtTarget.restrict hf _) (IsLocalAtTarget.restrict hg _) U.2
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))] at hg
+    obtain ⟨hY, hg⟩ := hg
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))] at hf
+    obtain ⟨hX, hf⟩ := hf
+    rw [HasAffineProperty.iff_of_isAffine (P := P) (Q := (affineAnd Q))]
+    exact ⟨hX, hQ _ _ hg hf⟩
+
+/-- If `P` is a morphism property affine locally defined by `affineAnd Q`, `P` is stable under
+base change if `Q` is. -/
+lemma HasAffineProperty.affineAnd_stableUnderBaseChange {P : MorphismProperty Scheme.{u}}
+    [HasAffineProperty P (affineAnd Q)] (hQi : RingHom.RespectsIso Q)
```

Comment:
I think this HasAffineProperty should not be an instance argument and a regular argument instead since we will not have these instances lying around.